### ISSUE TITLE
`bundle update foo` can not find compatible versions

### ISSUE THREAD

**Issue by marcandre** [bundler/bundler#5095]
I may be missing something, but my understanding is that if `bundle install` works fine, then a subsequent `bundle update foo` should never fail  (assuming no changes to the `Gemfile`). At worst it should resolve `foo` to whatever version is already in the `Gemfile.lock`, no?

I've attached my `Gemfile` & `Gemfile.lock`. To reproduce my issue, `bundle && bundle update ember-rails`. For some reason it's trying to upgrade Rails to v5 and (rightfully) failing:

```
$ bundle -v
Bundler version 1.13.5
$ bundle
[...]
Bundle complete! 193 Gemfile dependencies, 389 gems now installed [...]
$ bundle update ember-rails
[...]Resolving dependencies.............................................................
Bundler could not find compatible versions for gem "activesupport":
  In Gemfile:
    active_model_serializers (~> 0.8.0) was resolved to 0.8.3, which depends on
      activemodel (>= 3.0) was resolved to 5.0.0.1, which depends on
        activesupport (= 5.0.0.1)

    jquery-rails was resolved to 1.0.1, which depends on
      railties (~> 3.0) was resolved to 3.0.0, which depends on
        activesupport (= 3.0.0)
Bundler could not find compatible versions for gem "rack":
  In Gemfile:
    rack (~> 1.6)

    jquery-rails was resolved to 4.0.1, which depends on
      railties (>= 4.2.0.beta) was resolved to 5.0.0.1, which depends on
        actionpack (= 5.0.0.1) was resolved to 5.0.0.1, which depends on
          rack (~> 2.0)
Bundler could not find compatible versions for gem "rails":
  In snapshot (Gemfile.lock):
    rails (= 4.2.7)

  In Gemfile:
    rails (~> 4.2.7)

    jquery-rails was resolved to 0.1.1, which depends on
      rails (~> 3.0.0.rc)
```

[Gemfile_with_lock.zip](https://github.com/bundler/bundler/files/533835/Gemfile_with_lock.zip)

PS: I've made no attempt to minimize my `Gemfile` to get that effect, let me know if I should.

**Comment by segiddins** [comment]
I can reproduce this, but have no idea why it's happening. A minimal reproduction would definitely help pinpoint what exactly in bundler is causing the issue

**Comment by segiddins** [comment]
Worth noting this is a regression from Bundler 1.11.2

**Comment by segiddins** [comment]
I have a sneaky fix locally for this

**Comment by marcandre** [comment]
I'm not sure what the PR does exactly, but I fear it doesn't really fix this.

With this particular Gemfile, the correct fix would upgrade `ember-rails` to '0.18.5'; newer versions require an upgrade to `ember-data-source` which is locked. I arrived to this conclusion by manually trying different requirements and doing `bundle`.

Is your patch arriving to the same result, or is it keeping it at '0.16.4'?

**Comment by segiddins** [comment]
Updates ember-rails to 0.18.5

**Comment by marcandre** [comment]
Oh, great!

I note that it updaded `jquery-rails` and `execjs` needlessly (currently locked versions are ok), but it is already much better than failing to upgrade :smile:. Might be a different bug altogether with the new conservative upgrading, I can open another issue afterwards.

**Comment by segiddins** [comment]
Nope, since those aren't explicitly included in the Gemfile, bundler is free to update them

**Comment by marcandre** [comment]
I'm quite surprised that explicit inclusion in Gemfile makes a difference, I don't quite see why that's the case and would lean to disagree with that.
In any case, `execjs` isn't listed in the `Gemfile`, but `jquery-rails` is...

**Comment by chrismo** [comment]
I can try to explain the explicit inclusion in Gemfile (I need to sorta write this up for related work anyway, so this was a good excuse):

If a given Gemfile and lockfile haven't changed, they've been pulled fresh from git, then `bundle install` of course should just install the same versions as listed in the lockfile.

If a change is made to the Gemfile, say a newer version of gem `foo`, then `bundle install` is run, then that gem is unlocked and free to move. Bundler also needs to decide what to do about all of `foo`'s dependencies. 

If you only asked for `gem 'foo'` in your Gemfile and `foo` depends on `bar` and `qux` - when you want `foo` to move you shouldn't have to also tell Bundler to update its dependencies, you probably don't care or want to care about its dependencies, just get on with it. So Bundler will treat it all as a unit and unlock them all and allow them all to move.

The problem comes when _other_ gems you request in your Gemfile might also depend on `bar` or `qux`. Now Bundler has a dilemma: you want `foo` to move, and `foo` needs `bar` and `qux`, but you still have `tam` in your Gemfile, and it's not moving, and it also depends on `qux`. 

If Bundler chooses to favor `foo` by allowing `qux` to move, it could break `tam` altogether. So Bundler chooses to let `foo` and `bar` be unlocked and move, but keeps `qux` locked. This could also lead to a problem, but it will at least report a VersionConflict involving `foo`, the changed gem, instead of `tam`, the gem you probably didn't know also depended on `qux`. 

So, this leads us to the rule: "All of `foo`'s dependencies will also be unlocked and be free to move IF they also aren't dependencies of a different declared gem in the Gemfile."

But what if one of `foo`'s dependencies is itself a _declared_ dependency in the Gemfile? For example, your Gemfile is `gem 'foo'` and `gem 'bar'`. Well, really it's essentially the same rule as before, just that `bar` isn't a dependency of another declared gem, it IS a declared gem.

This is (hopefully) what [this section on Conservative Updates](http://bundler.io/v1.12/man/bundle-install.1.html#CONSERVATIVE-UPDATING) in `bundle install` docs describes. 

Now to bring this back around to _this_ ticket: `bundle update` has no such conservative update option for _shared_ dependencies, though the _declared_ gem version of the rule still holds. 

If you think `bundle update` _should_ have a similar conservative update option for shared dependencies, it's almost done: https://github.com/bundler/bundler/pull/4980.

**Comment by miks** [comment]
If I understand correctly, problem mentioned here is fixed with 1.14.4.
But I'm still getting weird errors like this:
```
    releaf was resolved to 1.0.8, which depends on
      releaf-core (= 1.0.8) was resolved to 1.0.8, which depends on
        railties (~> 4.2.7.1) was resolved to 4.2.7.1, which depends on
          actionpack (= 4.2.7.1) was resolved to 4.2.7.1, which depends on
            activesupport (= 5.0.1)

    releaf was resolved to 1.0.8, which depends on
      releaf-core (= 1.0.8) was resolved to 1.0.8, which depends on
        railties (~> 4.2.7.1) was resolved to 4.2.7.1, which depends on
          activesupport (= 4.2.7.1)
```
Where in fact  actionpack 4.2.7.1 depends on activesupport 4.2.7.1  (https://rubygems.org/gems/actionpack/versions/4.2.7.1)

bundler 1.13.6 is working correctly. I got this error on travis-ci where latest (1.14.4) version is used.

**Comment by segiddins** [comment]
@miks please open a new issue with a gemfile we can use to debug, thanks!
